File: app/src/main/java/com/bubelov/coins/ui/activity/EditPlaceActivity.java
package com.bubelov.coins.ui.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Button;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import com.bubelov.coins.R;
import com.bubelov.coins.repository.place.PlacesRepository;
import com.bubelov.coins.model.Place;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.OnClick;

/**
 * @author Igor Bubelov
 */

public class EditPlaceActivity extends AbstractActivity implements OnMapReadyCallback {
    private static final String ID_EXTRA = "id";

    private static final String MAP_CAMERA_POSITION_EXTRA = "map_camera_position";

    private static final int REQUEST_PICK_LOCATION = 10;

    private static final float DEFAULT_ZOOM = 13;

    @BindView(R.id.closed_switch)
    Switch closedSwitch;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.name)
    TextView name;

    @BindView(R.id.change_location)
    Button changeLocation;

    @BindView(R.id.phone)
    TextView phone;

    @BindView(R.id.website)
    TextView website;

    @BindView(R.id.description)
    TextView description;

    @BindView(R.id.opening_hours)
    TextView openingHours;

    @Inject
    PlacesRepository placesRepository;

    private Place place;

    private GoogleMap map;

    private LatLng pickedLocation;

    public static void startForResult(Activity activity, long placeId, CameraPosition mapCameraPosition, int requestCode) {
        Intent intent = new Intent(activity, EditPlaceActivity.class);
        intent.putExtra(ID_EXTRA, placeId);
        intent.putExtra(MAP_CAMERA_POSITION_EXTRA, mapCameraPosition);
        activity.startActivityForResult(intent, requestCode, ActivityOptionsCompat.makeSceneTransitionAnimation(activity).toBundle());
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dependencies().inject(this);
        setContentView(R.layout.activity_edit_place);
        ButterKnife.bind(this);

        toolbar.setNavigationOnClickListener(view -> supportFinishAfterTransition());
        toolbar.inflateMenu(R.menu.edit_place);

        toolbar.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.action_send) {
                if (place == null) {
                    if (name.length() == 0) {
                        showAlert(R.string.name_is_not_specified);
                        return true;
                    }

                    if (pickedLocation == null) {
                        showAlert(R.string.location_is_not_specified);
                        return true;
                    }
                }

                if (place == null) {
                    new AddPlaceTask().execute();
                    return true;
                } else {
                    new UpdatePlaceTask().execute();
                    return true;
                }
            }

            return false;
        });

        place = placesRepository.getPlace(getIntent().getLongExtra(ID_EXTRA, -1));

        if (place == null) {
            toolbar.setTitle(R.string.action_add_place);
            closedSwitch.setVisibility(View.GONE);
            changeLocation.setText(R.string.set_location);
        } else {
            name.setText(place.name());
            changeLocation.setText(R.string.change_location);
            phone.setText(place.phone());
            website.setText(place.website());
            description.setText(place.description());
            openingHours.setText(place.openingHours());
        }

        SupportMapFragment mapFragment = (SupportMapFragment)getSupportFragmentManager().findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_PICK_LOCATION && resultCode == RESULT_OK) {
            pickedLocation = data.getParcelableExtra(PickLocationActivity.LOCATION_EXTRA);
            map.moveCamera(CameraUpdateFactory.newLatLngZoom(pickedLocation, DEFAULT_ZOOM));
            changeLocation.setText(R.string.change_location);
        }

        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        map = googleMap;

        map.getUiSettings().setAllGesturesEnabled(false);

        if (place != null) {
            map.moveCamera(CameraUpdateFactory.newLatLngZoom(place.getPosition(), DEFAULT_ZOOM));
        }
    }

    private Place getEditedPlace() {
        return Place.builder()
                .id(place == null ? 0 : place.id())
                .name(name.getText().toString())
                .description(description.getText().toString())
                .latitude(pickedLocation.latitude)
                .longitude(pickedLocation.longitude)
                .categoryId(place == null ? 0 : place.categoryId())
                .phone(phone.getText().toString())
                .website(website.getText().toString())
                .openingHours(openingHours.getText().toString())
                .address(place == null ? "" : place.address())
                .visible(!closedSwitch.isChecked())
                .build();
    }

    @OnCheckedChanged(R.id.closed_switch)
    public void onClosedSwitchChanged() {
        boolean closed = closedSwitch.isChecked();

        name.setEnabled(!closed);
        phone.setEnabled(!closed);
        website.setEnabled(!closed);
        description.setEnabled(!closed);
        openingHours.setEnabled(!closed);
    }

    @OnClick(R.id.change_location)
    public void onChangeLocationClick() {
        PickLocationActivity.startForResult(this, place == null ? null : place.getPosition(), getIntent().getParcelableExtra(MAP_CAMERA_POSITION_EXTRA), REQUEST_PICK_LOCATION);
    }

    private class AddPlaceTask extends AsyncTask<Void, Void, Boolean> {
        @Override
        protected void onPreExecute() {
            showProgress();
        }

        @Override
        protected Boolean doInBackground(Void... args) {
            return placesRepository.add(getEditedPlace());
        }

        @Override
        protected void onPostExecute(Boolean success) {
            hideProgress();

            if (success) {
                setResult(RESULT_OK);
                supportFinishAfterTransition();
            } else {
                Toast.makeText(EditPlaceActivity.this, R.string.could_not_add_place, Toast.LENGTH_LONG).show();
            }
        }
    }

    private class UpdatePlaceTask extends AsyncTask<Void, Void, Boolean> {
        @Override
        protected void onPreExecute() {
            showProgress();
        }

        @Override
        protected Boolean doInBackground(Void... voids) {
            return placesRepository.update(getEditedPlace());
        }

        @Override
        protected void onPostExecute(Boolean success) {
            hideProgress();

            if (success) {
                setResult(RESULT_OK);
                supportFinishAfterTransition();
            } else {
                Toast.makeText(EditPlaceActivity.this, R.string.could_not_update_place, Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/com/bubelov/coins/ui/activity/EditPlaceActivity.kt
/*
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

package com.bubelov.coins.ui.activity

import android.app.Activity
import android.arch.lifecycle.*
import android.arch.lifecycle.Observer
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.view.View
import com.bubelov.coins.BuildConfig

import com.bubelov.coins.R
import com.bubelov.coins.model.Location
import com.bubelov.coins.model.Place
import com.bubelov.coins.ui.viewmodel.EditPlaceViewModel
import com.bubelov.coins.util.toLatLng
import com.bubelov.coins.util.toLocation
import com.bubelov.coins.util.viewModelProvider
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions

import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.activity_edit_place.*
import javax.inject.Inject

class EditPlaceActivity : DaggerAppCompatActivity() {
    @Inject lateinit var modelFactory: ViewModelProvider.Factory
    private val model by lazy { viewModelProvider(modelFactory) as EditPlaceViewModel }

    private val map = MutableLiveData<GoogleMap>()
    private var placeLocationMarker: Marker? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_place)

        model.init(intent.getSerializableExtra(PLACE_EXTRA) as Place)

        toolbar.setNavigationOnClickListener { supportFinishAfterTransition() }
        toolbar.inflateMenu(R.menu.edit_place)

        toolbar.setOnMenuItemClickListener { item ->
            if (item.itemId == R.id.action_send) {
                submit()
            }

            true
        }

        val place = model.place

        if (place.id == 0L) {
            toolbar.setTitle(R.string.action_add_place)
            closed_switch.visibility = View.GONE
        } else {
            name.setText(place.name)
            phone.setText(place.phone)
            website.setText(place.website)
            description.setText(place.description)
            opening_hours.setText(place.openingHours)
        }

        (fragmentManager.findFragmentById(R.id.map) as MapFragment).getMapAsync({
            map.value = it
        })

        closed_switch.setOnCheckedChangeListener { _, checked ->
            name.isEnabled = !checked
            change_location.visibility = if (checked) View.GONE else View.VISIBLE
            phone.isEnabled = !checked
            website.isEnabled = !checked
            description.isEnabled = !checked
            opening_hours.isEnabled = !checked
        }

        change_location.setOnClickListener {
            val intent = PickLocationActivity.newIntent(
                this,
                LatLng(place.latitude, place.longitude).toLocation(),
                map.value?.cameraPosition?.zoom ?: DEFAULT_ZOOM
            )

            startActivityForResult(intent, REQUEST_PICK_LOCATION)
        }

        model.showProgress.observe(this, Observer {
            if (isFinishing) {
                return@Observer
            }

            state_switcher.displayedChild = if (it == true) 1 else 0
        })

        map.observe(this, Observer { map ->
            if (map == null) return@Observer

            map.uiSettings.setAllGesturesEnabled(false)

            setMarker(map, LatLng(place.latitude, place.longitude).toLocation())

            map.moveCamera(
                CameraUpdateFactory.newLatLngZoom(
                    LatLng(place.latitude, place.longitude),
                    DEFAULT_ZOOM
                )
            )
        })

        model.submittedSuccessfully.observe(this, Observer {
            when (it) {
                true -> {
                    setResult(Activity.RESULT_OK)
                    finish()
                }

                false -> {
                    AlertDialog.Builder(this)
                        .setMessage(R.string.could_not_submit_changes)
                        .setPositiveButton(android.R.string.ok, null)
                        .show()
                }
            }
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_PICK_LOCATION && resultCode == Activity.RESULT_OK && data != null) {
            val map = map.value
            val location =
                data.getSerializableExtra(PickLocationActivity.LOCATION_EXTRA) as Location

            if (map != null) {
                setMarker(map, location)

                map.moveCamera(
                    CameraUpdateFactory.newLatLngZoom(
                        location.toLatLng(),
                        DEFAULT_ZOOM
                    )
                )
            }
        }

        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun setMarker(map: GoogleMap, location: Location) {
        placeLocationMarker?.remove()

        placeLocationMarker = map.addMarker(
            MarkerOptions()
                .position(location.toLatLng())
                .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_map_marker_empty))
                .anchor(BuildConfig.MAP_MARKER_ANCHOR_U, BuildConfig.MAP_MARKER_ANCHOR_V)
        )
    }

    private fun submit() {
        syncUIWithModel()

        if (name.length() == 0) {
            AlertDialog.Builder(this)
                .setMessage(R.string.name_is_not_specified)
                .setPositiveButton(android.R.string.ok, null)
                .show()
            return
        }

        model.submitChanges()
    }

    private fun syncUIWithModel() {
        model.place = model.place.copy(
            visible = !closed_switch.isChecked,
            name = name.text.toString(),
            description = description.text.toString(),
            latitude = map.value?.cameraPosition?.target?.latitude ?: 0.0,
            longitude = map.value?.cameraPosition?.target?.longitude ?: 0.0,
            phone = phone.text.toString(),
            website = website.text.toString(),
            openingHours = opening_hours.text.toString()
        )
    }

    companion object {
        private const val PLACE_EXTRA = "place"

        private const val REQUEST_PICK_LOCATION = 10

        private const val DEFAULT_ZOOM = 13f

        fun startForResult(
            activity: Activity,
            place: Place,
            requestCode: Int
        ) {
            val intent = Intent(activity, EditPlaceActivity::class.java)
            intent.putExtra(PLACE_EXTRA, place)
            activity.startActivityForResult(intent, requestCode)
        }
    }
}